crates/chia-sdk-driver/src/primitives/mips/memo.rs

```diff
@@ -0,0 +1,572 @@
+
+            if restriction.curry_tree_hash() == self.puzzle_hash.into() {
+                return Some(ParsedRestriction::EnforceDelegatedPuzzleWrappers(
+                    restriction,
+                    items.iter().map(|item| item.memo).collect(),
+                ));
+            }
+        }
+
+        if let Ok(memo) = Force1of2RestrictedVariableMemo::from_clvm(allocator, self.memo) {
+            let restriction = Force1of2RestrictedVariable::new(
+                memo.left_side_subtree_hash,
+                memo.nonce,
+                memo.member_validator_list_hash,
+                memo.delegated_puzzle_validator_list_hash,
+            );
+
+            if restriction.curry_tree_hash() == self.puzzle_hash.into() {
+                return Some(ParsedRestriction::Force1of2RestrictedVariable(restriction));
+            }
+        }
+
+        if let Ok(seconds) = Option::<u64>::from_clvm(allocator, self.memo) {
+            for &seconds in seconds.iter().chain(ctx.timelocks.iter()) {
+                let restriction = Timelock::new(seconds);
+
+                if restriction.curry_tree_hash() == self.puzzle_hash.into() {
+                    return Some(ParsedRestriction::Timelock(restriction));
+                }
+            }
+        }
+
+        None
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(list)]
+pub struct WrapperMemo<T = NodePtr> {
+    pub puzzle_hash: Bytes32,
+    pub memo: T,
+}
+
+impl WrapperMemo<NodePtr> {
+    pub fn new(puzzle_hash: Bytes32, memo: NodePtr) -> Self {
+        Self { puzzle_hash, memo }
+    }
+
+    pub fn force_assert_coin_announcement() -> Self {
+        Self {
+            puzzle_hash: FORCE_ASSERT_COIN_ANNOUNCEMENT_PUZZLE_HASH.into(),
+            memo: NodePtr::NIL,
+        }
+    }
+
+    pub fn force_coin_message() -> Self {
+        Self {
+            puzzle_hash: FORCE_COIN_MESSAGE_PUZZLE_HASH.into(),
+            memo: NodePtr::NIL,
+        }
+    }
+
+    pub fn prevent_multiple_create_coins() -> Self {
+        Self {
+            puzzle_hash: PREVENT_MULTIPLE_CREATE_COINS_PUZZLE_HASH.into(),
+            memo: NodePtr::NIL,
+        }
+    }
+
+    pub fn timelock(
+        allocator: &mut Allocator,
+        seconds: u64,
+        reveal: bool,
+    ) -> Result<Self, DriverError> {
+        Ok(Self {
+            puzzle_hash: Timelock::new(seconds).curry_tree_hash().into(),
+            memo: if reveal {
+                seconds.to_clvm(allocator)?
+            } else {
+                NodePtr::NIL
+            },
+        })
+    }
+
+    pub fn prevent_condition_opcode(
+        allocator: &mut Allocator,
+        opcode: u16,
+        reveal: bool,
+    ) -> Result<Self, DriverError> {
+        Ok(Self {
+            puzzle_hash: PreventConditionOpcode::new(opcode).curry_tree_hash().into(),
+            memo: if reveal {
+                opcode.to_clvm(allocator)?
```

Comment:
are we encoding features of the puzzle in the memo? I expect that we still validate those features by actually running the puzzle, right?